File: backend/classes/Invoice.php
<?php
class Invoice {
    private $conn;
    private $table_name = "invoices";

    public function __construct($db) {
        $this->conn = $db;
    }

    public function getAll($company_id) {
        $query = "SELECT i.*, c.name as client_name
                  FROM " . $this->table_name . " i
                  LEFT JOIN clients c ON i.client_id = c.id
                  WHERE i.company_id = ?
                  ORDER BY i.created_at DESC";

        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(1, $company_id, PDO::PARAM_INT);
        $stmt->execute();

        $invoices = [];
        while ($row = $stmt->fetch(PDO::FETCH_ASSOC)) {
            $invoice_data = $row;
            $invoice_data['items'] = $this->getInvoiceItems($row['id']);
            $invoices[] = $invoice_data;
        }

        return $invoices;
    }

    public function getById($id, $company_id) {
        $query = "SELECT i.*, c.name as client_name
                  FROM " . $this->table_name . " i
                  LEFT JOIN clients c ON i.client_id = c.id
                  WHERE i.id = ? AND c.company_id = ?";

        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(1, $id, PDO::PARAM_INT);
        $stmt->bindParam(2, $company_id, PDO::PARAM_INT);
        $stmt->execute();

        $invoice = $stmt->fetch(PDO::FETCH_ASSOC);
        if ($invoice) {
            $invoice['items'] = $this->getInvoiceItems($id);
            return $invoice;
        }

        return null;
    }

    private function getInvoiceItems($invoice_id) {
        // The items are stored as JSON in the items column of the invoices table
        $query = "SELECT items FROM " . $this->table_name . " WHERE id = ?";

        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(1, $invoice_id, PDO::PARAM_INT);
        $stmt->execute();

        $result = $stmt->fetch(PDO::FETCH_ASSOC);
        if ($result && $result['items']) {
            $items_data = json_decode($result['items'], true);
            if (is_array($items_data)) {
                return $items_data;
            }
        }

        return [];
    }

    public function create($data) {
        // Generate invoice number if not provided
        $invoice_number = $data['invoice_number'] ?? $this->generateInvoiceNumber($data['company_id']);

        $query = "INSERT INTO " . $this->table_name . " 
                  (company_id, client_id, quote_id, invoice_number, due_date, status, subtotal, tax_amount, total, paid_amount, notes, items)
                  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";

        $stmt = $this->conn->prepare($query);
        
        $company_id = $data['company_id'];
        $client_id = $data['client_id'];
        $quote_id = $data['quote_id'] ?? null;
        $due_date = $data['due_date'] ?? null;
        $status = $data['status'] ?? 'draft';
        $subtotal = $data['subtotal'];
        $tax_amount = $data['tax_amount'];
        $total = $data['total'] ?? $data['total_amount'] ?? 0; // Handle both field names
        $paid_amount = $data['paid_amount'] ?? 0;
        $notes = $data['notes'] ?? null;
        $items_json = isset($data['items']) ? json_encode($data['items']) : '[]';

        $stmt->bindParam(1, $company_id, PDO::PARAM_INT);
        $stmt->bindParam(2, $client_id, PDO::PARAM_INT);
        $stmt->bindParam(3, $quote_id, PDO::PARAM_INT);
        $stmt->bindParam(4, $invoice_number, PDO::PARAM_STR);
        $stmt->bindParam(5, $due_date, PDO::PARAM_STR);
        $stmt->bindParam(6, $status, PDO::PARAM_STR);
        $stmt->bindParam(7, $subtotal, PDO::PARAM_STR);
        $stmt->bindParam(8, $tax_amount, PDO::PARAM_STR);
        $stmt->bindParam(9, $total, PDO::PARAM_STR);
        $stmt->bindParam(10, $paid_amount, PDO::PARAM_STR);
        $stmt->bindParam(11, $notes, PDO::PARAM_STR);
        $stmt->bindParam(12, $items_json, PDO::PARAM_STR);

        if ($stmt->execute()) {
            $invoice_id = $this->conn->lastInsertId();
            return $this->getById($invoice_id, $company_id);
        }

        return null;
    }

    public function createFromQuote($quote_id, $company_id) {
        // Get the quote data
        $quote_query = "SELECT q.*, c.name as client_name
                        FROM quotes q
                        LEFT JOIN clients c ON q.client_id = c.id
                        WHERE q.id = ? AND q.company_id = ?";

        try {
            $stmt = $this->conn->prepare($quote_query);
            $stmt->bindParam(1, $quote_id, PDO::PARAM_INT);
            $stmt->bindParam(2, $company_id, PDO::PARAM_INT);
            $stmt->execute();

            $quote = $stmt->fetch(PDO::FETCH_ASSOC);
            if (!$quote) {
                return null;
            }

            // Get quote items
            $items_query = "SELECT qi.*, i.name as item_name, i.description
                            FROM quote_items qi
                            LEFT JOIN items i ON qi.item_id = i.id
                            WHERE qi.quote_id = ?";

            $stmt = $this->conn->prepare($items_query);
            $stmt->bindParam(1, $quote_id, PDO::PARAM_INT);
            $stmt->execute();

            $quote_items = [];
            $subtotal = 0;
            while ($item = $stmt->fetch(PDO::FETCH_ASSOC)) {
                $total_price = $item['unit_price'] * $item['quantity'];
                $subtotal += $total_price;
                
                $quote_items[] = [
                    'item_id' => $item['item_id'],
                    'item_name' => $item['item_name'] ?? 'Custom Item',
                    'description' => $item['description'] ?? '',
                    'unit_price' => $item['unit_price'],
                    'quantity' => $item['quantity'],
                    'total_price' => $total_price
                ];
            }

            // Calculate tax (assuming 10% for now, this could be configurable)
            $tax_rate = 0.10;
            $tax_amount = $subtotal * $tax_rate;
            $total_amount = $subtotal + $tax_amount;

            // Calculate due date (30 days from today)
            $due_date = date('Y-m-d', strtotime('+30 days'));

            // Create invoice data
            $invoice_data = [
                'company_id' => $company_id,
                'client_id' => $quote['client_id'],
                'quote_id' => $quote_id,
                'due_date' => $due_date,
                'status' => 'draft',
                'subtotal' => $subtotal,
                'tax_amount' => $tax_amount,
                'total' => $total_amount,
                'paid_amount' => 0,
                'notes' => $quote['notes'],
                'items' => $quote_items
            ];

            return $this->create($invoice_data);
        } catch (Exception $e) {
            error_log("Error in createFromQuote: " . $e->getMessage());
            throw $e;
        }
    }

    public function update($id, $data) {
        $query = "UPDATE " . $this->table_name . " 
                  SET client_id = ?, invoice_number = ?, due_date = ?, 
                      status = ?, subtotal = ?, tax_amount = ?, total = ?, 
                      paid_amount = ?, notes = ?, items = ?, updated_at = CURRENT_TIMESTAMP
                  WHERE id = ?";

        $stmt = $this->conn->prepare($query);
        
        $client_id = $data['client_id'];
        $invoice_number = $data['invoice_number'];
        $due_date = $data['due_date'] ?? null;
        $status = $data['status'];
        $subtotal = $data['subtotal'];
        $tax_amount = $data['tax_amount'];
        $total = $data['total'] ?? $data['total_amount'] ?? 0;
        $paid_amount = $data['paid_amount'] ?? 0;
        $notes = $data['notes'] ?? null;
        $items_json = isset($data['items']) ? json_encode($data['items']) : '[]';

        $stmt->bindParam(1, $client_id, PDO::PARAM_INT);
        $stmt->bindParam(2, $invoice_number, PDO::PARAM_STR);
        $stmt->bindParam(3, $due_date, PDO::PARAM_STR);
        $stmt->bindParam(4, $status, PDO::PARAM_STR);
        $stmt->bindParam(5, $subtotal, PDO::PARAM_STR);
        $stmt->bindParam(6, $tax_amount, PDO::PARAM_STR);
        $stmt->bindParam(7, $total, PDO::PARAM_STR);
        $stmt->bindParam(8, $paid_amount, PDO::PARAM_STR);
        $stmt->bindParam(9, $notes, PDO::PARAM_STR);
        $stmt->bindParam(10, $items_json, PDO::PARAM_STR);
        $stmt->bindParam(11, $id, PDO::PARAM_INT);

        if ($stmt->execute()) {
            return $this->getById($id, $data['company_id']);
        }

        return null;
    }

    public function delete($id, $company_id) {
        // Verify the invoice belongs to the company
        $verify_query = "SELECT i.id FROM " . $this->table_name . " i
                         LEFT JOIN clients c ON i.client_id = c.id
                         WHERE i.id = ? AND c.company_id = ?";
        
        $stmt = $this->conn->prepare($verify_query);
        $stmt->bindParam(1, $id, PDO::PARAM_INT);
        $stmt->bindParam(2, $company_id, PDO::PARAM_INT);
        $stmt->execute();

        $result = $stmt->fetch(PDO::FETCH_ASSOC);
        if (!$result) {
            return false;
        }

        $query = "DELETE FROM " . $this->table_name . " WHERE id = ?";
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(1, $id, PDO::PARAM_INT);

        return $stmt->execute();
    }

    private function generateInvoiceNumber($company_id) {
        $query = "SELECT COUNT(*) as count FROM " . $this->table_name . " i
                  LEFT JOIN clients c ON i.client_id = c.id
                  WHERE c.company_id = ?";
        
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(1, $company_id, PDO::PARAM_INT);
        $stmt->execute();
        $row = $stmt->fetch(PDO::FETCH_ASSOC);
        
        $count = $row['count'] + 1;
        return 'INV-' . date('Y') . '-' . str_pad($count, 4, '0', STR_PAD_LEFT);
    }
}
?>

File: lib/models/invoice.dart
import 'dart:convert';

class InvoiceItem {
  final int? id;
  final int? invoiceId;
  final int? itemId;
  final String itemName;
  final String description;
  final double unitPrice;
  final double quantity;
  final double totalPrice;

  InvoiceItem({
    this.id,
    this.invoiceId,
    this.itemId,
    required this.itemName,
    required this.description,
    required this.unitPrice,
    required this.quantity,
    required this.totalPrice,
  });

  factory InvoiceItem.fromMap(Map<String, dynamic> map) {
    return InvoiceItem(
      id: map['id'] is String ? int.parse(map['id']) : map['id'],
      invoiceId: map['invoice_id'] is String ? int.parse(map['invoice_id']) : map['invoice_id'],
      itemId: map['item_id'] is String ? int.parse(map['item_id']) : map['item_id'],
      itemName: map['item_name'] ?? '',
      description: map['description'] ?? '',
      unitPrice: map['unit_price'] is String ? double.parse(map['unit_price']) : (map['unit_price'] ?? 0.0).toDouble(),
      quantity: map['quantity'] is String ? double.parse(map['quantity']) : (map['quantity'] ?? 1.0).toDouble(),
      totalPrice: map['total_price'] is String ? double.parse(map['total_price']) : (map['total_price'] ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'invoice_id': invoiceId,
      'item_id': itemId,
      'item_name': itemName,
      'description': description,
      'unit_price': unitPrice,
      'quantity': quantity,
      'total_price': totalPrice,
    };
  }

  Map<String, dynamic> toApiMap() {
    return {
      'item_id': itemId?.toString(),
      'item_name': itemName,
      'description': description,
      'unit_price': unitPrice.toString(),
      'quantity': quantity.toString(),
      'total_price': totalPrice.toString(),
    };
  }

  InvoiceItem copyWith({
    int? id,
    int? invoiceId,
    int? itemId,
    String? itemName,
    String? description,
    double? unitPrice,
    double? quantity,
    double? totalPrice,
  }) {
    return InvoiceItem(
      id: id ?? this.id,
      invoiceId: invoiceId ?? this.invoiceId,
      itemId: itemId ?? this.itemId,
      itemName: itemName ?? this.itemName,
      description: description ?? this.description,
      unitPrice: unitPrice ?? this.unitPrice,
      quantity: quantity ?? this.quantity,
      totalPrice: totalPrice ?? this.totalPrice,
    );
  }

  @override
  String toString() {
    return 'InvoiceItem{itemName: $itemName, quantity: $quantity, unitPrice: $unitPrice, totalPrice: $totalPrice}';
  }
}

class Invoice {
  final int? id;
  final int companyId;
  final int clientId;
  final int? quoteId; // Reference to original quote if converted
  final String invoiceNumber;
  final DateTime invoiceDate;
  final DateTime? dueDate;
  final String status; // unpaid, partial, paid, cancelled
  final double subtotal;
  final double taxAmount;
  final double totalAmount;
  final double paidAmount;
  final String? notes;
  final List<InvoiceItem> items;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Invoice({
    this.id,
    required this.companyId,
    required this.clientId,
    this.quoteId,
    required this.invoiceNumber,
    required this.invoiceDate,
    this.dueDate,
    required this.status,
    required this.subtotal,
    required this.taxAmount,
    required this.totalAmount,
    this.paidAmount = 0.0,
    this.notes,
    required this.items,
    this.createdAt,
    this.updatedAt,
  });

  double get remainingAmount => totalAmount - paidAmount;
  bool get isOverdue => dueDate != null && DateTime.now().isAfter(dueDate!) && status != 'paid';
  bool get isPaid => status == 'paid';
  bool get isPartiallyPaid => paidAmount > 0 && paidAmount < totalAmount && status != 'paid';

  factory Invoice.fromMap(Map<String, dynamic> map) {
    
    // Parse invoice items if they exist
    List<InvoiceItem> invoiceItems = [];
    if (map['items'] != null) {
      if (map['items'] is List) {
        invoiceItems = (map['items'] as List).map((item) => InvoiceItem.fromMap(item)).toList();
      } else if (map['items'] is String) {
        // Handle case where items might be stored as JSON string
        try {
          final itemsJson = json.decode(map['items']);
          if (itemsJson is List) {
            invoiceItems = itemsJson.map((item) => InvoiceItem.fromMap(item)).toList();
          }
        } catch (e) {
          // Silently handle JSON parsing errors
        }
      }
    }
    
    return Invoice(
      id: map['id'] is String ? int.parse(map['id']) : map['id'],
      companyId: map['company_id'] is String ? int.parse(map['company_id']) : map['company_id'],
      clientId: map['client_id'] is String ? int.parse(map['client_id']) : map['client_id'],
      quoteId: map['quote_id'] != null 
          ? (map['quote_id'] is String ? int.parse(map['quote_id']) : map['quote_id'])
          : null,
      invoiceNumber: map['invoice_number'] ?? '',
      invoiceDate: map['invoice_date'] != null 
          ? DateTime.parse(map['invoice_date']) 
          : DateTime.now(), // Default to current date if not provided
      dueDate: map['due_date'] != null ? DateTime.parse(map['due_date']) : null,
      status: map['status'] ?? 'draft',
      subtotal: map['subtotal'] is String ? double.parse(map['subtotal']) : (map['subtotal'] ?? 0.0).toDouble(),
      taxAmount: map['tax_amount'] is String ? double.parse(map['tax_amount']) : (map['tax_amount'] ?? 0.0).toDouble(),
      totalAmount: map['total_amount'] != null 
          ? (map['total_amount'] is String ? double.parse(map['total_amount']) : map['total_amount'].toDouble())
          : (map['total'] != null 
              ? (map['total'] is String ? double.parse(map['total']) : map['total'].toDouble())
              : 0.0), // Handle both 'total_amount' and 'total' field names
      paidAmount: map['paid_amount'] is String ? double.parse(map['paid_amount']) : (map['paid_amount'] ?? 0.0).toDouble(),
      notes: map['notes'],
      items: invoiceItems,
      createdAt: map['created_at'] != null ? DateTime.parse(map['created_at']) : null,
      updatedAt: map['updated_at'] != null ? DateTime.parse(map['updated_at']) : null,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'company_id': companyId,
      'client_id': clientId,
      'quote_id': quoteId,
      'invoice_number': invoiceNumber,
      'invoice_date': invoiceDate.toIso8601String(),
      'due_date': dueDate?.toIso8601String(),
      'status': status,
      'subtotal': subtotal,
      'tax_amount': taxAmount,
      'total_amount': totalAmount,
      'paid_amount': paidAmount,
      'notes': notes,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  Map<String, dynamic> toApiMap() {
    return {
      'client_id': clientId.toString(),
      'quote_id': quoteId?.toString(),
      'invoice_number': invoiceNumber,
      'invoice_date': invoiceDate.toIso8601String().split('T')[0], // YYYY-MM-DD format
      'due_date': dueDate?.toIso8601String().split('T')[0],
      'status': status,
      'subtotal': subtotal.toString(),
      'tax_amount': taxAmount.toString(),
      'total_amount': totalAmount.toString(),
      'paid_amount': paidAmount.toString(),
      'notes': notes ?? '',
      'items': items.map((item) => item.toApiMap()).toList(),
    };
  }

  Invoice copyWith({
    int? id,
    int? companyId,
    int? clientId,
    int? quoteId,
    String? invoiceNumber,
    DateTime? invoiceDate,
    DateTime? dueDate,
    String? status,
    double? subtotal,
    double? taxAmount,
    double? totalAmount,
    double? paidAmount,
    String? notes,
    List<InvoiceItem>? items,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Invoice(
      id: id ?? this.id,
      companyId: companyId ?? this.companyId,
      clientId: clientId ?? this.clientId,
      quoteId: quoteId ?? this.quoteId,
      invoiceNumber: invoiceNumber ?? this.invoiceNumber,
      invoiceDate: invoiceDate ?? this.invoiceDate,
      dueDate: dueDate ?? this.dueDate,
      status: status ?? this.status,
      subtotal: subtotal ?? this.subtotal,
      taxAmount: taxAmount ?? this.taxAmount,
      totalAmount: totalAmount ?? this.totalAmount,
      paidAmount: paidAmount ?? this.paidAmount,
      notes: notes ?? this.notes,
      items: items ?? this.items,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  String toString() {
    return 'Invoice{id: $id, invoiceNumber: $invoiceNumber, status: $status, totalAmount: $totalAmount}';
  }
}